### PR TITLE
Do not use wrappers for circular evaluation tracking

### LINKED ISSUE #27695: Android Studio project import memory regression in 8.6-rc1
### Current Behavior

There seems to be a memory regression in 8.6-rc1 during project configuration.

Our 2000 sub-project benchmark for Android Studio import reports significantly more memory use after project evaluation in Gradle 8.6-rc1

image

### Expected Behavior

See image above for before/after values.

### Context (optional)

_No response_

### Steps to Reproduce

1. Clone the benchmark project. `git clone https://github.com/gavra0/android-extra-large-benchmark -b 2022.9`
2. Optional but recommended, set up a smaller subset (i.e. maybe 200):  https://github.com/gavra0/android-extra-large-benchmark/tree/2022.9?tab=readme-ov-file#working-with-smaller-subsets
3. Enable the [measurement plugin](https://github.com/gavra0/android-extra-large-benchmark/blob/2022.9/measurements/plugin.init.gradle.kts) for use during sync: https://github.com/gavra0/android-extra-large-benchmark/tree/2022.9?tab=readme-ov-file#working-with-smaller-subsets
4. Open Android Studio and import the project
5. Once import finishes, captured histograms will be found at `measurements` directory.

### Gradle version

8.6-rc1

### Gradle version that used to work

8.4

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

### PR DESCRIPTION
**Pull request by mlopatkin**
Fixes #27695.

### Context
Apparently, providers are responsible for the significant part of the build's memory, and increasing their footprint by even a small margin is very noticeable. This PR replaces wrapper-based approach with explicit scope management.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by mlopatkin** [comment]
@bot-gradle test RfN, APT

**Comment by mlopatkin** [comment]
@bot-gradle merge

**Comment by mlopatkin** [comment]
@bot-gradle merge

**Comment by cobexer** [comment]
@bot-gradle merge

File: platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java [hunk 1]
```diff
@@ -41,37 +41,42 @@ protected String toStringNoReentrance() {
 
     @Override
     public boolean calculatePresence(ValueConsumer consumer) {
-        if (!left.calculatePresence(consumer) || !right.calculatePresence(consumer)) {
-            return false;
+        try (EvaluationContext.ScopeContext ignored = openScope()) {
+            if (!left.calculatePresence(consumer) || !right.calculatePresence(consumer)) {
```
Comment (bamboo) [inline platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java:45]: ❓ Shouldn't the context be passed down to both `calculatePresence` calls here?

Comment (mlopatkin) [reply platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java:45]: > ❓ Shouldn't the context be passed down to both `calculatePresence` calls here?

No, not in the current implementation. The context is being passed around (as a token) when we're calling methods of the provider being currently evaluated, to highlight two things:
- `this` shouldn't be added to the context again
- we can safely evaluate "inner" providers/user code inside the method.

I'm thinking about alternative designs where providers can indeed receive context and register themselves (saving us a ThreadLocal lookup), but these ideas don't tick all the boxes yet to ensure that all Provider implementations are forced to follow the protocol.

Comment (bamboo) [reply platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java:45]: I see.

**Pull request merged**